### ISSUE TITLE
Concurrent EVP_PKEY_pairwise_check calls seem to cause a deadlock [FIPS enabled]

### ISSUE THREAD

**Issue by shaan1337** [openssl/openssl#20825]
Hello,

I think that I'm running into a deadlock in the FIPS module when there are multiple simultaneous calls to `EVP_PKEY_pairwise_check`
I can easily reproduce it but I don't have a minimal repro (i can try if needed).

I've attached the managed (.NET) and native LLDB stack traces for the same run: [stack_traces.zip](https://github.com/openssl/openssl/files/11321878/stack_traces.zip)

There are several threads stuck, apparently waiting for a lock to be released:
```
  thread #56, name = '.NET ThreadPool', stop reason = signal SIGSTOP
    frame #0: 0x00007f03a68ac319 libc.so.6`__futex_abstimed_wait_common + 73
    frame #1: 0x00007f03a68b5d7f libc.so.6`pthread_rwlock_wrlock@@GLIBC_2.34 + 303
    frame #2: 0x00007f02bef440bf fips.so`ossl_drbg_lock + 31
    frame #3: 0x00007f02bef42869 fips.so`ossl_drbg_lock_parent + 25
    frame #4: 0x00007f02bef461fa fips.so`get_parent_reseed_count.lto_priv.0 + 122
    frame #5: 0x00007f02bef47af3 fips.so`ossl_prov_drbg_generate.part.0 + 467
    frame #6: 0x00007f02bef344bc fips.so`EVP_RAND_generate + 300
    frame #7: 0x00007f02bef563a9 fips.so`bnrand.constprop.0 + 329
    frame #8: 0x00007f02bef56525 fips.so`bnrand_range.constprop.0 + 213
    frame #9: 0x00007f02bef52672 fips.so`ossl_bn_miller_rabin_is_prime.part.0 + 1042
    frame #10: 0x00007f02bef560b0 fips.so`bn_is_prime_int.part.0.constprop.0 + 480
    frame #11: 0x00007f02bef4492a fips.so`ossl_rsa_check_prime_factor + 42
    frame #12: 0x00007f02beef79f9 fips.so`rsa_validate.lto_priv.0 + 1257
    frame #13: 0x00007f02e42df9f1 libcrypto.so.3`try_provided_check + 97
    frame #14: 0x00007f02e42dfd25 libcrypto.so.3`EVP_PKEY_pairwise_check + 37
    frame #15: 0x00007f03a4129180 libSystem.Security.Cryptography.Native.OpenSsl.so`CryptoNative_DecodePkcs8PrivateKey + 128
```

System specs:
openssl version: `OpenSSL 3.0.1 14 Dec 2021 (Library: OpenSSL 3.0.1 14 Dec 2021)`
tested on: `RHEL 9 (with FIPS enabled)`

**Comment by t8m** [comment]
Can you reproduce this problem with newer 3.0.x release? There were locking fixes in later releases.

**Comment by shaan1337** [comment]
i've tested with openssl 3.0.8 (compiled from source) and ran the repro 10 times. the problem didn't occur but i think i'm missing something in the setup:

just to be sure i've also compiled openssl 3.0.1 (14 Dec 2021) from source and tested but the repro didn't work as well (i.e the deadlock didn't happen)

in both cases, i've compiled with `enable-fips enable-md2` and i can confirm that FIPS is enabled and working (tested with `openssl md5 /some/file`)

do you have any idea of what i could be missing?

**Comment by t8m** [comment]
It could be also some Red Hat specific patch in OpenSSL causing the problem.

**Comment by beldmit** [comment]
Could you please provide a reproducer? I'll try to reproduce it on Red Hat.

**Comment by shaan1337** [comment]
@beldmit that would be great thanks.

i've cooked up a minimal repro by extracting the code for `CryptoNative_DecodePkcs8PrivateKey()`  from the `dotnet/runtime` repository then just called it from multiple threads and the deadlock occurs: https://github.com/shaan1337/openssl_deadlock_repro_20825 

I've not yet attempted to narrow down the exact method causing the problem.
It was tested on AWS with an official RHEL 9 AMI with FIPS enabled (runs `OpenSSL 3.0.1 14 Dec 2021`).

**Comment by neverpanic** [comment]
I ran your reproducer with the following patch applied (to see some output and fix the segfault at the end when entering a character) on RHEL 9.0 with `openssl-libs-3.0.1-46.el9_0.3.x86_64` in FIPS mode in the following configurations:

- `gcc -g -o openssl_deadlock_repro_20825 openssl_deadlock_repro_20825.c -lcrypto`
- `gcc -g -o openssl_deadlock_repro_20825 openssl_deadlock_repro_20825.c -lcrypto && valgrind --tool=helgrind ./openssl_deadlock_repro_20825`
- `clang -g -std=c99 -fsanitize=thread -o openssl_deadlock_repro_20825 openssl_deadlock_repro_20825.c -lcrypto`
- `gcc -g -std=c99 -fsanitize=thread -o openssl_deadlock_repro_20825 openssl_deadlock_repro_20825.c -lcrypto`

None of those configurations show the deadlock you saw. The output I added suggests the key was successfully decoded 10 times in all created threads. Can you provide more information on how you compile the reproducer, how you run it, and what version of OpenSSL is on your system (`rpm -q openssl-libs`)?

**Comment by shaan1337** [comment]
@neverpanic thank you for running the repro, i've used the following commands:
```
$ openssl genrsa -out rsa.key 2048
$ openssl pkcs8 -topk8 -inform PEM -in rsa.key -outform DER -out rsa.der -nocrypt
$ gcc -g openssl_deadlock_repro_20825.c -o openssl_deadlock_repro_20825 -lcrypto -lpthread -Wno-deprecated-declarations
$ ./openssl_deadlock_repro_20825
<deadlocks. verified with lldb which shows a stack trace similar to the one posted above>
```

The AMI was set up with default RHEL 9 on AWS (on a t3.xlarge instance) roughly as follows:
```
$ sudo fips-mode-setup --enable
$ sudo reboot

$ sudo yum install git
$ sudo yum install gcc
$ sudo yum install openssl-devel
```

Versions:
```
$ cat /etc/redhat-release 
Red Hat Enterprise Linux release 9.0 (Plow)

$ rpm -q openssl-libs
openssl-libs-3.0.1-47.el9_1.x86_64

$ rpm -q openssl
openssl-3.0.1-47.el9_1.x86_64

$ openssl version
OpenSSL 3.0.1 14 Dec 2021 (Library: OpenSSL 3.0.1 14 Dec 2021)
```

Can you please try increasing `nloops` to 100 to see if it makes any difference?

**Comment by neverpanic** [comment]
tl;dr: Install `openssl-3.0.7-6.el9_2` as soon as it is released.

I can't reproduce the deadlock, but with `openssl-libs-3.0.1-47.el9_1.x86_64`, I can reproduce what might look like a deadlock at a first glace. I see multiple threads waiting for the lock in `fips.so:ossl_drbg_lock_parent`, but at the same time, at least one of the threads is in `libgcrypt.so.3:ossl_rand_get_entropy`, `libgcrypt.so.3:ossl_pool_acquire_entropy` and `lib.so.6:getrandom`:

Thread in `getrandom(2)`:
```
  thread #9, name = 'openssl_deadloc', stop reason = signal SIGSTOP
    frame #0: 0x00007fcfdf4d1aaf libc.so.6`getrandom + 79
    frame #1: 0x00007fcfdf8f99d0 libcrypto.so.3`ossl_pool_acquire_entropy + 96
    frame #2: 0x00007fcfdf85bca1 libcrypto.so.3`ossl_rand_get_entropy + 49
    frame #3: 0x00007fcfdc3acd87 fips.so`crngt_get_entropy.lto_priv.0 + 87
    frame #4: 0x00007fcfdc39e6a1 fips.so`get_entropy.lto_priv.0 + 1025
    frame #5: 0x00007fcfdc3ad4e6 fips.so`ossl_prov_drbg_reseed.constprop.0 + 262
    frame #6: 0x00007fcfdc39fb13 fips.so`ossl_prov_drbg_generate.part.0 + 499
    frame #7: 0x00007fcfdc39fd38 fips.so`ossl_drbg_get_seed + 184
    frame #8: 0x00007fcfdc39e348 fips.so`get_entropy.lto_priv.0 + 168
    frame #9: 0x00007fcfdc3ad4e6 fips.so`ossl_prov_drbg_reseed.constprop.0 + 262
    frame #10: 0x00007fcfdc39fb13 fips.so`ossl_prov_drbg_generate.part.0 + 499
    frame #11: 0x00007fcfdc38c4bc fips.so`EVP_RAND_generate + 300
    frame #12: 0x00007fcfdc3ae3a9 fips.so`bnrand.constprop.0 + 329
    frame #13: 0x00007fcfdc3ae525 fips.so`bnrand_range.constprop.0 + 213
    frame #14: 0x00007fcfdc3aa672 fips.so`ossl_bn_miller_rabin_is_prime.part.0 + 1042
    frame #15: 0x00007fcfdc3ae0b0 fips.so`bn_is_prime_int.part.0.constprop.0 + 480
    frame #16: 0x00007fcfdc39c92a fips.so`ossl_rsa_check_prime_factor + 42
```

Thread waiting for the lock:
```
  thread #7, name = 'openssl_deadloc', stop reason = signal SIGSTOP
    frame #0: 0x00007fcfdf515319 libc.so.6`__futex_abstimed_wait_common + 73
    frame #1: 0x00007fcfdf51ed7f libc.so.6`pthread_rwlock_wrlock@@GLIBC_2.34 + 303
    frame #2: 0x00007fcfdc39c0bf fips.so`ossl_drbg_lock + 31
    frame #3: 0x00007fcfdc39a869 fips.so`ossl_drbg_lock_parent + 25
    frame #4: 0x00007fcfdc39e1fa fips.so`get_parent_reseed_count.lto_priv.0 + 122
    frame #5: 0x00007fcfdc39faf3 fips.so`ossl_prov_drbg_generate.part.0 + 467
    frame #6: 0x00007fcfdc38c4bc fips.so`EVP_RAND_generate + 300
    frame #7: 0x00007fcfdc3ae3a9 fips.so`bnrand.constprop.0 + 329
    frame #8: 0x00007fcfdc3ae525 fips.so`bnrand_range.constprop.0 + 213
    frame #9: 0x00007fcfdc3aa672 fips.so`ossl_bn_miller_rabin_is_prime.part.0 + 1042
    frame #10: 0x00007fcfdc3ae0b0 fips.so`bn_is_prime_int.part.0.constprop.0 + 480
    frame #11: 0x00007fcfdc39c92a fips.so`ossl_rsa_check_prime_factor + 42
    frame #12: 0x00007fcfdc34f9f9 fips.so`rsa_validate.lto_priv.0 + 1257
    frame #13: 0x00007fcfdf822af1 libcrypto.so.3`try_provided_check + 97
    frame #14: 0x00007fcfdf822e25 libcrypto.so.3`EVP_PKEY_pairwise_check + 37
```

If you run your reproducer under `strace`, e.g. using `strace -f -e getrandom ./openssl_deadlock_repro_20825`, you'll see that it is actually progressing, but constantly re-seeding the DRBGs from the OS using `getrandom(2)` calls:

```
getrandom("\x55\xb1\x95\x61\x80\x60\x8f\xc8", 8, GRND_NONBLOCK) = 8
strace: Process 14476 attached
strace: Process 14477 attached
strace: Process 14478 attached
strace: Process 14479 attached
strace: Process 14480 attached
strace: Process 14481 attached
strace: Process 14482 attached
strace: Process 14483 attached
[pid 14479] getrandom("\xe9\x6a\x7c\x49\xb7\xdb\x93\x8d\x46\xb3\x38\x81\xee\xba\xb1\x31", 16, 0) = 16
[pid 14479] getrandom("\x6d\x79\xf8\x6f\x72\x18\x89\xbb\x9b\x47\x6c\xe2\x87\xb3\x6c\xb1", 16, 0) = 16
[pid 14479] getrandom("\xbe\x43\xc4\xa6\xf1\xb0\x78\x69\x22\x1c\xa8\x27\x5d\xac\x5b\xe4", 16, 0) = 16
[pid 14479] getrandom("\x5a\x51\x80\x63\x5e\x85\xcf\x3b\xa8\xe1\x1d\x46\x90\xe7\xff\xa8", 16, 0) = 16
[pid 14479] getrandom("\x53\xba\xf4\xc8\x67\x79\xd1\x4d\xa8\x75\x2c\x6b\x21\xf1\x00\xa8", 16, 0) = 16
[pid 14479] getrandom("\x90\x27\xe5\x34\xbb\xd2\xd1\xce\xcf\xe9\x62\xa4\x0c\x7d\x75\x3e", 16, 0) = 16
[pid 14479] getrandom("\x91\x23\xc9\x23\xff\x37\xf5\x3b\xd1\x30\xf2\xa5\x70\x93\xb4\x80", 16, 0) = 16
[pid 14480] getrandom("\x4b\x8e\x49\x26\xa9\xcb\xdf\x30\x0e\x70\xb3\xd7\xa6\x5c\x6f\x5a", 16, GRND_RANDOM) = 16
[pid 14480] getrandom("\xb0\xa6\x9d\x2a\xd5\x03\xa0\x09\x83\x38\x02\xcf\x06\x67\xf5\x16", 16, GRND_RANDOM) = 16
[pid 14480] getrandom("\x55\xdb\xaf\xad\xab\xf7\x16\xda\xb3\xc9\x59\xf5\x5f\xc4\x3f\xbc", 16, GRND_RANDOM) = 16
[pid 14477] getrandom("\x07\x42\x02\xe2\xe7\x8d\xb6\x28\xba\x46\xae\xbe\x56\x35\x62\x00", 16, GRND_RANDOM) = 16
[pid 14477] getrandom("\x49\xa5\x82\x0f\x63\x32\xea\x08\x4d\xa2\x5e\x30\x74\x89\x5a\x57", 16, GRND_RANDOM) = 16
[pid 14477] getrandom("\xb5\x67\x2e\xf7\x6b\x1c\xe3\x8e\xbd\x0f\xe4\xe9\x19\x5a\x37\xc9", 16, GRND_RANDOM) = 16
[pid 14480] getrandom("\x77\xec\xca\x28\xbf\x48\x8c\xf2\xf5\x29\x1f\x23\x10\xd0\x3e\x53", 16, GRND_RANDOM) = 16
[pid 14480] getrandom("\xfc\x50\xf4\xaf\xae\xd8\x79\xee\xbb\xeb\xef\xd4\x8c\x63\x0b\x56", 16, GRND_RANDOM) = 16
[pid 14480] getrandom("\xaf\x56\x13\xf0\xd5\x7f\x22\xa6\x6c\x45\x6b\x87\x09\xcc\xc7\xb2", 16, GRND_RANDOM) = 16
[pid 14477] getrandom("\x35\xfb\x74\x4a\xc5\x12\xa6\x6e\x9d\xdf\x38\xc7\x57\x55\x7c\x93", 16, GRND_RANDOM) = 16
[pid 14477] getrandom("\xdc\x87\x39\xcd\x1c\x9f\xaf\x16\x2c\xe9\x3e\x62\x36\xe0\x9c\x3a", 16, GRND_RANDOM) = 16
[pid 14477] getrandom("\x18\x2c\x3e\x4a\xa5\x58\x6e\xb6\x88\x50\xc6\xef\x27\xfc\xe9\xef", 16, GRND_RANDOM) = 16
[pid 14479] getrandom("\xb9\xc8\x13\xd2\x03\x83\x06\xc2\x90\x87\x5f\x91\x82\x76\x35\x07", 16, GRND_RANDOM) = 16
[pid 14479] getrandom("\xdf\xef\x7d\xd1\xc4\x4a\x48\xc5\x90\xce\x44\x4d\x2f\x8e\xa6\x6d", 16, GRND_RANDOM) = 16
[pid 14479] getrandom("\x75\x22\x85\xed\x39\xc0\xa5\x88\x15\xb9\xa2\x8b\x6a\xf3\xd4\x10", 16, GRND_RANDOM) = 16
[pid 14482] getrandom("\x9b\x07\x37\x80\xad\x8b\x3f\x2d\xd9\x30\x15\x9b\xc0\x10\x39\x5e", 16, GRND_RANDOM) = 16
[pid 14482] getrandom("\x1c\x60\x90\xcd\xea\xef\xb8\xa7\xb0\xef\x53\x81\x0b\x9a\x61\xc3", 16, GRND_RANDOM) = 16
[pid 14482] getrandom("\x61\x6e\xb5\x1a\xdc\xe0\x76\x85\x3c\xbe\xe4\x42\x4f\xf0\xdc\x6d", 16, GRND_RANDOM) = 16
[pid 14483] getrandom("\xc3\x64\xad\xf7\xa2\xdd\xd4\x8c\xf8\xd4\x9d\x4f\xb9\xc7\xeb\x18", 16, GRND_RANDOM) = 16
[pid 14483] getrandom("\xa5\xcd\xc6\xcc\xf1\x69\xd8\xd2\x32\x82\xaa\xed\x2c\xc5\xdf\x47", 16, GRND_RANDOM) = 16
[pid 14483] getrandom("\x59\xca\xf6\xa6\x36\x00\xb5\x2d\xda\x4b\xeb\xc8\x79\x00\x89\x69", 16, GRND_RANDOM) = 16
[pid 14477] getrandom("\xcd\x3f\x8e\x45\x91\xf9\xc3\xf9\x41\xf5\xf6\x44\xaa\xcf\x96\x88", 16, GRND_RANDOM) = 16
[pid 14477] getrandom("\x79\xfd\x1d\xca\x3e\x6a\x94\x02\xe2\x92\x62\xea\x72\x79\xb6\x0e", 16, GRND_RANDOM) = 16
[pid 14477] getrandom("\x08\xb3\x02\x9e\xa5\x37\x09\x65\x64\x03\xc4\x5b\x69\x9e\x5b\x0c", 16, GRND_RANDOM) = 16
[pid 14479] getrandom("\x76\x6e\x6f\x90\x6f\xcf\xcc\x7b\x7f\x06\x4b\x60\xdd\xd8\x0f\x65", 16, GRND_RANDOM) = 16
[pid 14479] getrandom("\x12\xd6\x5b\x82\x98\xdd\x99\x86\xc0\xe4\x7b\x97\xf4\x7e\xc7\xa0", 16, GRND_RANDOM) = 16
...
```

This will eventually finish, but it will take a very long time.

While working on FIPS certification, Red Hat has received instructions from its certification lab that due to the use of chained DRBGs in OpenSSL, these chained DRBGs should pull randomness from their parents with prediction resistance enabled. Because each thread has its own DRBG, their first reseed causes a reseed of all parent DRBGs. Next, all child DRBGs in every child reseed whenever the parent DRBG has been reseeded. This causes every single draw from the DRBG to reseed from the OS. For your case, that means every single draw from a DRBG probably causes multiple calls to `getrandom(2)`, which is what slows your example down.

There is some good news: (a) this isn't a problem in OpenSSL upstream, so this ticket can technically by closed. (b) this is fixed in RHEL 9.2 (https://bugzilla.redhat.com/show_bug.cgi?id=2168224) and 9.0.z (https://bugzilla.redhat.com/show_bug.cgi?id=2178054). You are using 9.1, which will be EOL soon (I believe the 9.2 release date is today or tomorrow, which will obsolete 9.1). The release of 9.2 is also why this wasn't fixed in 9.1.z. That's also why I couldn't reproduce this, since I was testing `openssl-libs-3.0.1-46.el9_0.3`, which has the fix.

See also https://github.com/openssl/openssl/issues/20414 for previous discussion of the same issue. Specifically, see https://github.com/openssl/openssl/issues/20414#issuecomment-1450324003, https://github.com/openssl/openssl/issues/20414#issuecomment-1451479226 and https://github.com/openssl/openssl/issues/20414#issuecomment-1453196027 (the latter also explains the use of the `GRND_RANDOM` flag).

Note that the same situation will happen in OpenSSL without Red Hat patches when the DRBG is used with prediction resistance. https://github.com/openssl/openssl/commit/d90bd3468a9a8d2af6b821be50c1034e21d782ca documents this.

**Comment by shaan1337** [comment]
@neverpanic hey, thanks very much for your thorough investigation and response - it clearly explains the issue.

i can confirm that i have also seen the library calls to acquire entropy and getrandom(). I had some doubts that it could just be slow but the locks in the stack trace gave the impression that it was a deadlock.

i'm closing the issue, thanks!

**Comment by mattcaswell** [comment]
>     frame #2: 0x00007fcfdc39c0bf fips.so`ossl_drbg_lock + 31
>    frame #3: 0x00007fcfdc39a869 fips.so`ossl_drbg_lock_parent + 25
>    frame #4: 0x00007fcfdc39e1fa fips.so`get_parent_reseed_count.lto_priv.0 + 122

This looks like the issue I have proposed a fix for in #20944.

**Comment by beldmit** [comment]
@mattcaswell as @neverpanic mentioned, it's partially caused by Red Hat downstream patches causing reseeding.
